tests/math/test_trace_distance_math.py

```diff
@@ -0,0 +1,288 @@
+            np.sqrt(2) / 2,
+        ),
+        (
+            (
+                [[0.5, 0], [0, 0.5]],
+                [[[1, 0], [0, 0]], [[0, 0], [0, 1]], [[0.2, 0], [0, 0.8]]],
+            ),
+            [0.5, 0.5, 0.3],
+        ),
+        # Batch-Batch-TD
+        (([[[1, 0], [0, 0]]], [[[0.5, 0], [0, 0.5]]]), 0.5),
+        (([[[0, 0], [0, 1]]], [[[0.5, 0], [0, 0.5]]]), 0.5),
+        (([[[1, 0], [0, 0]]], [[[0.5, 0.5], [0.5, 0.5]]]), np.sqrt(2) / 2),
+        (([[[0, 0], [0, 1]]], [[[0.5, 0.5], [0.5, 0.5]]]), np.sqrt(2) / 2),
+        (
+            (
+                [[[0.5, 0, 0, 0.5], [0, 0, 0, 0], [0, 0, 0, 0], [0.5, 0, 0, 0.5]]],
+                [
+                    [
+                        [0.25, 0.25, 0.25, 0.25],
+                        [0.25, 0.25, 0.25, 0.25],
+                        [0.25, 0.25, 0.25, 0.25],
+                        [0.25, 0.25, 0.25, 0.25],
+                    ]
+                ],
+            ),
+            np.sqrt(2) / 2,
+        ),
+        (
+            (
+                [[[0.5, 0], [0, 0.5]]],
+                [[[1, 0], [0, 0]], [[0, 0], [0, 1]], [[0.2, 0], [0, 0.8]]],
+            ),
+            [0.5, 0.5, 0.3],
+        ),
+        (
+            (
+                [[[0.5, 0], [0, 0.5]], [[0, 0], [0, 1]], [[0.1, 0], [0, 0.9]]],
+                [[[1, 0], [0, 0]], [[0, 0], [0, 1]], [[0.2, 0], [0, 0.8]]],
+            ),
+            [0.5, 0, 0.1],
+        ),
+    ]
+
+    array_funcs = [
+        lambda x: x,
+        onp.array,
+        np.array,
+        jnp.array,
+        torch.tensor,
+        tf.Variable,
+        tf.constant,
+    ]
+
+    check_state = [True, False]
+
+    @pytest.mark.parametrize("states,td", state0_state1_td)
+    @pytest.mark.parametrize("check_state", check_state)
+    @pytest.mark.parametrize("func", array_funcs)
+    def test_trace_distance(self, states, check_state, td, func):
+        """Test trace distance between different quantum states."""
+        state0, state1 = states
+        state0 = func(state0)
+        state1 = func(state1)
+
+        assert qml.math.allclose(td, qml.math.trace_distance(state0, state1, check_state))
+        assert qml.math.allclose(td, qml.math.trace_distance(state1, state0, check_state))
+
+    d_mat_wrong_shape = [
+        # Shape that is not a power of 2
+        (([[1, 0, 0], [0, 0, 0], [0, 0, 0]], [[1, 0], [0, 0]])),
+        (([[1, 0], [0, 0]], [[1, 0, 0], [0, 0, 0], [0, 0, 0]])),
+        (([[[1, 0, 0], [0, 0, 0], [0, 0, 0]]], [[1, 0], [0, 0]])),
+        (([[1, 0], [0, 0]], [[[1, 0, 0], [0, 0, 0], [0, 0, 0]]])),
+        # Len of shape that is not in (2, 3)
+        (([1, 0], [[1, 0], [0, 0]])),
+        (([[1, 0], [0, 0]], [1, 0])),
+        # Shapes with different dimensions
+        (([[1, 0, 0], [0, 0, 0]], [[1, 0], [0, 0]])),
+        (([[1, 0], [0, 0]], [[1, 0, 0], [0, 0, 0]])),
+        (([[[1, 0, 0], [0, 0, 0]]], [[1, 0], [0, 0]])),
+        (([[1, 0], [0, 0]], [[[1, 0, 0], [0, 0, 0]]])),
+    ]
```

Comment:
It seems we have a few redundant tests here, since in `test_density_matrix_wrong_shape` we also swap the two positions.